text/3642-thread-spawn-hook.md

```diff
@@ -0,0 +1,238 @@
+unit test in a thread local variable, such that multiple tests can be run in
+parallel in the same process.
+
+However, this information will not be preserved across threads when a unit test
+will spawn a new thread, which is problematic.
+
+The solution seems to be "inheriting thread locals": thread locals that are
+automatically inherited by new threads.
+
+However, adding this property to thread local variables is not easily possible.
+Thread locals are initialized lazily. And by the time they are initialized, the
+parent thread might have already disappeared, such that there is no value left
+to inherit from.
+Additionally, even if the parent thread was still alive, there is no way to
+access the value in the parent thread without causing race conditions.
+
+Allowing hooks to be run as part of spawning a thread allows precise control
+over how thread locals are "inherited".
+One could simply `clone()` them, but one could also add additional information
+to them, or even add relevant information to some (global) data structure.
+
+For example, not only could a custom testing framework keep track of unit test
+state even across spawned threads, but a logging/debugging/tracing library could
+keeps track of which thread spawned which thread to provide more useful
+information to the user.
+
+# Public Interface
+
+```rust
+// In std::thread:
+
+/// Registers a function to run for every new thread spawned.
+///
+/// The hook is executed in the parent thread, and returns a function
+/// that will be executed in the new thread.
+///
+/// The hook is called with the `Thread` handle for the new thread.
+///
+/// If the hook returns an `Err`, thread spawning is aborted. In that case, the
+/// function used to spawn the thread (e.g. `std::thread::spawn`) will return
+/// the error returned by the hook.
+///
+/// Hooks can only be added, not removed.
+///
+/// The hooks will run in order, starting with the most recently added.
+///
+/// # Usage
+///
+/// ```
+/// std::thread::add_spawn_hook(|_| {
+///     ..; // This will run in the parent (spawning) thread.
+///     Ok(move || {
+///         ..; // This will run it the child (spawned) thread.
+///     })
+/// });
+/// ```
+///
+/// # Example
+///
+/// ```
+/// thread_local! {
+///     static MY_THREAD_LOCAL: Cell<u32> = Cell::new(0);
+/// }
+///
+/// std::thread::add_spawn_hook(|_| {
+///     // Get the value of X in the spawning thread.
+///     let value = MY_THREAD_LOCAL.get();
+///
+///     Ok(move || {
+///         // Set the value of X in the newly spawned thread.
+///         MY_THREAD_LOCAL.set(value);
+///     })
+/// });
+/// ```
+pub fn add_spawn_hook<F, G>(hook: F)
+where
+    F: 'static + Sync + Fn(&Thread) -> std::io::Result<G>,
+    G: 'static + Send + FnOnce();
+```
+
+# Implementation
+
+The implementation could simply be a static `RwLock` with a `Vec` of
+(boxed/leaked) `dyn Fn`s, or a simple lock free linked list of hooks.
```

Comment:
As an alternative, we could use [global registration](https://github.com/rust-lang/rust/issues/125119)
- Limits what you can do in your `hook`
- Would require a macro and some complexity
- Would only work for static linking with `std`
- Blocks this on a language feature proposal

Reply:
Are there any advantages to that approach? I can only think of downsides ^^'.

Reply:
Removes all overhead when not used and keeps the overhead minimal.  Likely won't be a problem.  I mostly mentioned this because I had the thought to use this when it was posted but as I wrote up my post I realized how long the downsides were and thought it worth recording in case others have the thought.

Reply:
FWIW, the cost of spawning a new thread (not only allocating its stack, its Thread object, etc., but also the actual creation of the thread through a syscall) completely dwarfs any overhead from an indirect function call, so I don't think overhead is an argument here.

(For similar reasons, `std::panic::set_hook` is also just a simple function, rather than a magical attribute like no_std's `#[panic_handler]`.)

I'll add a note in the RFC on this.